flow/layers/picture_layer.cc

```diff
@@ -17,6 +18,68 @@ PictureLayer::PictureLayer(const SkPoint& offset,
       is_complex_(is_complex),
       will_change_(will_change) {}
 
+void PictureLayer::Diff(DiffContext* context, const Layer* old_layer) {
+  DiffContext::AutoSubtreeRestore subtree(context);
+  if (!context->IsSubtreeDirty()) {
+    assert(old_layer);
+    auto prev = old_layer->as_picture_layer();
+    if (!Compare(this, prev)) {
+      context->MarkSubtreeDirty(prev->paint_region());
+    }
+  }
+  context->PushTransform(SkMatrix::Translate(offset_.x(), offset_.y()));
+  context->AddPaintRegion(picture()->cullRect());
+  set_paint_region(context->CurrentSubtreeRegion());
+}
+
+bool PictureLayer::Compare(const PictureLayer* l1, const PictureLayer* l2) {
+  const auto& pic1 = l1->picture_.get();
+  const auto& pic2 = l2->picture_.get();
+  if (pic1.get() == pic2.get()) {
+    return true;
+  }
+  auto op_cnt_1 = pic1->approximateOpCount();
+  auto op_cnt_2 = pic2->approximateOpCount();
+  if (op_cnt_1 != op_cnt_2 || pic1->cullRect() != pic2->cullRect()) {
+    return false;
+  }
+
+  if (op_cnt_1 > 10) {
+    return false;
+  }
+
+  // TODO(knopp) we don't actually need the data; this could be done without
+  // allocations by implementing stream that calculates SHA hash and
+  // comparing those hashes
+  auto d1 = l1->SerializedPicture();
+  auto d2 = l2->SerializedPicture();
+  auto res = d1->equals(d2.get());
+  return res;
+}
+
+sk_sp<SkData> PictureLayer::SerializedPicture() const {
+  SkSerialProcs procs = {
+      nullptr,
+      nullptr,
+      [](SkImage* i, void* ctx) {
+        auto id = i->uniqueID();
+        return SkData::MakeWithCopy(&id, sizeof(id));
+      },
+      nullptr,
+      [](SkTypeface* tf, void* ctx) {
+        auto id = tf->uniqueID();
+        return SkData::MakeWithCopy(&id, sizeof(id));
+      },
+      nullptr,
+  };
+
+  if (!cached_serialized_picture_) {
+    cached_serialized_picture_ = picture_.get()->serialize(&procs);
```

Comment:
We don't need the procs if we don't enter this block so it should probably be moved in here.